accepted/00000-minion-selective-action-blocking.md

```diff
@@ -0,0 +1,139 @@
+
+For traditional clients the action types that are allowed when a system is locked are configured in the db. The configuration is global and applies to all clients indiscriminately.
+
+In the case of Salt minions, the same approach will be used. The action types that are allowed when a system is locked will be stored in the db.
+
+Additionally, in order to make the approach more flexible and to accommodate CaaSP a new object called a "locking profile" will be introduced. 
+
+The locking profile will contain a list of permitted actions. A profile can be reused across multiple minions. A minion will have only one locking profile assigned when locked. Specifying a locking profile will be optional when locking a minion. If not specified, the global locking configuration will apply.
+
+Current db schema:
+```
+        Table "rhnActionType"
+      Column      |          Type                                                                                         
+------------------+------------------------                                                                             
+ id               | numeric                
+ label            | character varying(48)  
+ name             | character varying(100) 
+ trigger_snapshot | character(1)                                                                                  
+ unlocked_only    | character(1)
+
+    Table "rhnserverlock"
+  Column   |           Type           
+-----------+--------------------------
+ server_id | numeric                  
+ locker_id | numeric                  
+ reason    | character varying(4000)  
+ created   | timestamp with time zone 
+
+```
+
+If column `unlocked_only` is `Y` the action type is blocked when the system is locked.
+
+Proposed enhancement with locking profiles:
+```
++         Table "rhnLockProfile"
+       Column      |          Type                                                                                         
+ ------------------+------------------------
+  id               | numeric
+  label            | character varying
+ 
++         Table "rhnLockProfileActionTypes"
+       Column      |          Type                                                                                         
+ ------------------+------------------------
+  profile_id       | numeric references rhnLockProfile(id)
+  action_type_id   | numeric references rhnActionType(id)
+  unlocked_only    | character(1)
+ 
+          Table "rhnServerLock"
+   Column    |           Type           
+ ------------+--------------------------
+  server_id  | numeric                  
+  locker_id  | numeric                  
+  reason     | character varying(4000)  
+  created    | timestamp with time zone 
++ profile_id | numeric nullable references rhnLockProfile(id)
+
+```
+The new column `rhnserverlock.profile_id` will reference the locking profile to be used. If it's `NULL` the global configuration from `rhnActionType.unlocked_only` will be used.
```

Comment:
What advantages or disadvantages you see with the choice of having such profiles defined in the database vs. in code?

Reply:
I think both approaches are valid. For the initial implementation we could maintain the profile in code. In the future, with a plugin system, the profile could be defined in code by the plugin. 
OTOH having the profile in the DB is consistent with the current implementation, it makes the profile easier to change and we could even create some UI/API to allow an admin user to define what actions are allowed on certain machines for certain users.

Reply:
My personal opinion is that we have far too many tables which are actually constants, and constants that the code needs to know about anyway, so in those cases they should not really be in the DB at all, following the principle that DBs are for variable data, not programming enums.

That said, I can see the case of having a table here if it is actually a good choice: fed via a product-specific plugin or maybe SCC (I would not particularly trust users for this use case). Until we have such variability, I would still recommend keeping this in code.

I agree with you this is inconsistent with some of the mechanisms we have elsewhere, but IMO we should rather refactor those and get rid of some tables (eg. `rhnActionStatus`, `rhnActionType`, `rhnChecksumType`...) - or at least avoid adding more. Having the same constant in the DB and code is just more maintenance burden without a clear advantage FMPOV.

Reply:
That's fine with me. Let's keep this in code for now. We can always add tables later if there's a need for it.

Reply:
While I agree with keeping with the principle that databases should contain variable data, aren't we creating a potential pitfall for managing the locking action in the future? Because if allowed actions for traditional systems will be in the database, whilst the allowed actions for minions will be in code.